domain/application/service/application.go

```diff
@@ -1481,3 +1534,153 @@ func (s *Service) ResolveControllerCharmDownload(ctx context.Context, resolve ap
+// application trust will be set to false.
+// If no application is found, an error satisfying
+// [applicationerrors.ApplicationNotFound] is returned.
+// If the charm does not exist, an error satisfying
+// [applicationerrors.CharmNotFound] is returned, if this is the case, then
+// the application is in a bad state and should be removed.
+// If the charm config is not valid, an error satisfying
+// [applicationerrors.CharmConfigNotValid] is returned.
+func (s *Service) SetApplicationConfig(ctx context.Context, appID coreapplication.ID, newConfig map[string]string) error {
+	if err := appID.Validate(); err != nil {
+		return internalerrors.Errorf("application ID: %w", err)
+	}
+
+	// Get the charm config. This should be safe to do outside of a singular
+	// transaction, as the charm config is immutable. So it will either be there
+	// or not, and if it's not there we can return an error stating that.
+	// Otherwise if it is there, but then is removed before we set the config, a
+	// foreign key constraint will be violated, and we can return that as an
+	// error.
+
+	// Return back the charm UUID, so that we can verify that the charm
+	// hasn't changed between this call and the transaction to set it.
+
+	charmID, cfg, err := s.st.GetCharmConfigByApplicationID(ctx, appID)
+	if err != nil {
+		return internalerrors.Capture(err)
+	}
+
+	charmConfig, err := decodeConfig(cfg)
+	if err != nil {
+		return internalerrors.Capture(err)
+	}
+
+	// Grab the application settings, which is currently just the trust setting.
+	trust, err := getTrustSettingFromConfig(newConfig)
+	if err != nil {
+		return internalerrors.Capture(err)
+	}
+
+	// Everything else from the newConfig is just application config. Treat it
+	// as such.
+	coercedConfig, err := charmConfig.ParseSettingsStrings(newConfig)
+	if errors.Is(err, internalcharm.ErrUnknownOption) {
+		return internalerrors.Errorf("%w: %w", applicationerrors.InvalidCharmConfig, err)
+	} else if err != nil {
+		return internalerrors.Capture(err)
+	}
+
+	// The encoded config is the application config, with the type of the
+	// option. Encoding the type ensures that if the type changes during an
+	// upgrade, we can prevent a runtime error during that phase.
+	encodedConfig := make(map[string]application.ApplicationConfig, len(coercedConfig))
+	for k, v := range coercedConfig {
+		option, ok := charmConfig.Options[k]
+		if !ok {
+			// This should never happen, as we've verified the config is valid.
+			// But if it does, then we should return an error.
+			return internalerrors.Errorf("missing charm config, expected %q", k)
+		}
+
+		optionType, err := encodeOptionType(option.Type)
+		if err != nil {
+			return internalerrors.Capture(err)
+		}
+
+		encodedConfig[k] = application.ApplicationConfig{
+			Value: v,
+			Type:  optionType,
+		}
+	}
+
+	return s.st.SetApplicationConfigAndSettings(ctx, appID, charmID, encodedConfig, application.ApplicationSettings{
```

Comment:
I know we already had this discussion multiple times, but in this particular case, isn't it concerning to have 2 separate tx? Can't this happen concurrently during deploy (if several units) for example?